Batching logic for same label set metrics

bryanlou opened the pull request:
**Description:**
This PR implements the necessary functionality to batch metrics with exact same label sets. Specifically, this PR accomplishes the following:

1. Create a new `GroupedMetric` struct in the metric_translator file which defines the Namespace, Labels, Metrics and Timestamp of the Opentelemetry metrics
2. Implement batching logic in EMF Exporter
3. Translate batched metrics to EMF logs 

**Testing:** 
This code change has been tested in a sample environment using an NGINX server on EKS with these settings in `config.yaml` 
```
exporters:
  logging:
    loglevel: debug
  awsemf:
    log_group_name: 'awscollector-test'
    region: 'us-west-2'
    log_stream_name: metric-declarations
    dimension_rollup_option: 'NoDimensionRollup'

service:
  pipelines:
    metrics:
      receivers: [prometheus]
      exporters: [awsemf, logging]

  extensions: [health_check, pprof, zpages]
```

Before, the CW logs look like this:
```
{
    “Namespace”: “kube-system”,
    “OTLib”: “Undefined”,
    “Service”: “kube-dns”,
    “_aws”: {
        “CloudWatchMetrics”: [
            {
                “Namespace”: “kubernetes-service-endpoints”,
                “Dimensions”: [
                    [
                        “Namespace”,
                        “Service”,
                        “container_name”,
                        “eks_amazonaws_com_component”,
                        “k8s_app”,
                        “kubernetes_io_cluster_service”,
                        “kubernetes_io_name”,
                        “kubernetes_node”,
                        “pod_name”,
                        “OTLib”
                    ]
                ],
                “Metrics”: [
                    {
                        “Name”: “go_goroutines”,
                        “Unit”: “”
                    }
                ]
            }
        ],
        “Timestamp”: 1603352885414
    },
    “container_name”: “coredns”,
    “eks_amazonaws_com_component”: “kube-dns”,
    “go_goroutines”: 0,
    “k8s_app”: “kube-dns”,
    “kubernetes_io_cluster_service”: “true”,
    “kubernetes_io_name”: “CoreDNS”,
    “kubernetes_node”: “ip-192-168-43-221.us-west-2.compute.internal”,
    “pod_name”: “coredns-5946c5d67c-txp4b”
}
```
And after batching, the newly exported CW logs:
```
{
    “Namespace”: “kube-system”,
    “OTLib”: “Undefined”,
    “Service”: “kube-dns”,
    “_aws”: {
        “CloudWatchMetrics”: [
            {
                “Namespace”: “kubernetes-service-endpoints”,
                “Dimensions”: [
                    [
                        “Namespace”,
                        “Service”,
                        “container_name”,
                        “eks_amazonaws_com_component”,
                        “k8s_app”,
                        “kubernetes_io_cluster_service”,
                        “kubernetes_io_name”,
                        “kubernetes_node”,
                        “pod_name”,
                        “OTLib”
                    ]
                ],
                “Metrics”: [
                    {
                        “Name”: “go_goroutines”,
                        “Unit”: “”
                    },
                    {
                        “Name”: “more_metrics”,
                        “Unit”: “”
                    },
                    {
                        .......
                    }
                ]
            }
        ],
        “Timestamp”: 1603352885414
    },
    “container_name”: “coredns”,
    “eks_amazonaws_com_component”: “kube-dns”,
    “go_goroutines”: 0,
    “k8s_app”: “kube-dns”,
    “kubernetes_io_cluster_service”: “true”,
    “kubernetes_io_name”: “CoreDNS”,
    “kubernetes_node”: “ip-192-168-43-221.us-west-2.compute.internal”,
    “pod_name”: “coredns-5946c5d67c-txp4b”
    “more_metrics”: “new_Metrics”
}
```

**Documentation**
Nothing in README has been updated as no changes were made to configurations or rules for exported metrics.

kohrapha reviewed (commented):
I'm wondering if your output is correct. Should the exported API be changed? For example, `Dimensions` is now a KV map instead of a 2D array, and `Metrics` has a different structure as well. 

@hdj630 could you clarify if this is the correct output?

kohrapha commented:
Could you pull in the latest changes from `master` and resolve the conflicts?

kohrapha commented on exporter/awsemfexporter/factory.go:
```diff
@@ -16,7 +16,6 @@ package awsemfexporter
 
 import (
 	"context"
-
```
nit: could you remove all these whitespace changes as well?

kohrapha commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -116,12 +124,73 @@ func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, dimensionRollupOption stri
 	return cwMetricLists, totalDroppedMetrics
 }
 
+func batchCWMetrics(cwMetricList []*CWMetrics, groupedCWMetricMap map[string]*GroupedCWMetric, cwMetricsMap map[string]*CWMetrics) {
```
Can you add tests for this function?

kohrapha commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -116,12 +124,73 @@ func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, dimensionRollupOption stri
+	}
+}
+
+func TranslateBatchedMetricToEMF(groupedCWMetricMap map[string]*GroupedCWMetric) []*LogEvent {
```
Can you add tests for this function?

kohrapha commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -116,12 +124,73 @@ func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, dimensionRollupOption stri
 }
 
+func batchCWMetrics(cwMetricList []*CWMetrics, groupedCWMetricMap map[string]*GroupedCWMetric, cwMetricsMap map[string]*CWMetrics) {
+	for _, i := range cwMetricList {
```
nit: could we choose a better name than `i`? In this case, would `cwMetric` or `metric` be better?

kohrapha commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -116,12 +124,73 @@ func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, dimensionRollupOption stri
+func batchCWMetrics(cwMetricList []*CWMetrics, groupedCWMetricMap map[string]*GroupedCWMetric, cwMetricsMap map[string]*CWMetrics) {
+	for _, i := range cwMetricList {
+		if len(i.Measurements[0].Dimensions[0]) > 0 {
+			var key = strings.Join(i.Measurements[0].Dimensions[0],"")
```
is there a reason for the use of `var` here? If not, the standard is to use `:=` 
```suggestion
			key := strings.Join(i.Measurements[0].Dimensions[0], "")
```

kohrapha reviewed (commented):
Could you also run `make lint` to check formatting? 

@hdj630 @mxiamxia we should prob look into getting a CI pipeline for this repo to check linting and test cases

bryanlou commented:
@hdj630 @mxiamxia

mxiamxia commented:
From the example input/output. The dimensions are not in right format. `Dimensions` section should only list the dimension names and the key/value should be listed outside as we're having today before batched data.
```
"Dimensions": {
        "Namespace": "kube-system",
        "OTLib": "Undefined",
        "Service": "kube-dns",
        "container_name": "coredns",
        "eks_amazonaws_com_component": "kube-dns",
        "k8s_app": "kube-dns",
        "kubernetes_io_cluster_service": "true",
        "kubernetes_io_name": "CoreDNS",
        "kubernetes_node": "ip-192-168-43-221.us-west-2.compute.internal",
        "pod_name": "coredns-5946c5d67c-txp4b"
    },
```

kohrapha commented:
@mxiamxia I agree, I think the output here is not correct. And from what I understand of this CR, the goal was more of an internal-facing change; the external exported metrics/logs should not differ (except for an array of values vs single value).

bryanlou commented:
> @mxiamxia I agree, I think the output here is not correct. And from what I understand of this CR, the goal was more of an internal-facing change; the external exported metrics/logs should not differ (except for an array of values vs single value).

I fixed this, and updated the log format in the description

kohrapha commented:
@bryanlou can you merge in the changes from `master` and resolve those conflicts?

kohrapha commented:
@bryanlou did you merge correctly? Did you merge from upstream or from `master`? I see a lot of code changes in this CR that came from upstream and are not specific to your CR. Can you please fix them?

kohrapha commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -118,8 +127,8 @@ func (dps DoubleHistogramDataPointSlice) At(i int) DataPoint {
-func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, dimensionRollupOption string, namespace string) ([]*CWMetrics, int) {
-	var cwMetricList []*CWMetrics
+func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, dimensionRollupOption string, cwMetricsMap map[string]*CWMetrics, groupedCWMetricMap map[string]*GroupedCWMetric, namespace string) ([]*CWMetrics, int) {
+	var cwMetricLists []*CWMetrics
```
Don't think this line was meant to be changed? It should be `cwMetricList`.

kohrapha commented on exporter/awsemfexporter/metric_translator_test.go:
```diff
@@ -1323,14 +1323,161 @@ func BenchmarkTranslateOtToCWMetricWithNamespace(b *testing.B) {
 	}
 }
 
-func BenchmarkTranslateCWMetricToEMF(b *testing.B) {
```
Did you mean to remove this benchmark?

kohrapha commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -159,10 +168,101 @@ func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, dimensionRollupOption stri
 			cwMetrics := getCWMetrics(&metric, namespace, instrumentationLibName, dimensionRollupOption)
-			cwMetricList = append(cwMetricList, cwMetrics...)
+			cwMetricLists = append(cwMetricLists, cwMetrics...)
+			batchCWMetrics(cwMetrics, groupedCWMetricMap, cwMetricsMap)
```
Is this the best place to put the batching logic? It seems like we're performing redundant calculations of `createDimensions`. If we have a `groupedMetric` that share the same label set, we should perform dimension processing only once on them together. So, maybe it might be worthwhile taking a look at how we can place the batching logic either deeper within `buildCWMetric` or move the logic for dimension setting out of `getCWMetric` and have it perform on `groupedMetric` after batching.

kohrapha commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -159,10 +168,101 @@ func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, dimensionRollupOption stri
+
+func batchCWMetrics(cwMetricList []*CWMetrics, groupedCWMetricMap map[string]*GroupedCWMetric, cwMetricsMap map[string]*CWMetrics) {
+	for _, met := range cwMetricList {
+		if len(met.Measurements[0].Dimensions[0]) > 0 {
```
Is this error safe? What about the following edge cases:
1. `met.Measurements` is an empty array 
2. There is more than 1 element in `met.Measurements`
3. There is more than 1 dimension set

kohrapha commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -159,10 +168,101 @@ func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, dimensionRollupOption stri
+func batchCWMetrics(cwMetricList []*CWMetrics, groupedCWMetricMap map[string]*GroupedCWMetric, cwMetricsMap map[string]*CWMetrics) {
+	for _, met := range cwMetricList {
+		if len(met.Measurements[0].Dimensions[0]) > 0 {
+			key := strings.Join(met.Measurements[0].Dimensions[0], "")
```
I don't think you should use the dimensions as the key. Once we have the metric declaration logic, dimensions != labels. I think this ties in to my earlier comment of performing batching before dimension setting.

kohrapha commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -159,10 +168,101 @@ func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, dimensionRollupOption stri
+					groupedCWMetricMap[key].MetricUnits[v["Name"]] = v["Unit"]
+				}
+			} else {
+				cwMetricsMap[key]=met
```
nit: did you perform `go fmt` here?
```suggestion
				cwMetricsMap[key] = met
```

kohrapha commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -61,6 +62,14 @@ type CWMetrics struct {
+	Timestamp    int64
+	Dimensions   map[string]interface{}
+	MetricUnits  map[string]string
+}
```
From the design doc, should `GroupedMetric` only have `Timestamp`, `Fields`, and `Metrics` as attributes?

kohrapha commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -159,10 +168,101 @@ func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, dimensionRollupOption stri
+	return ples
 }
 
 func TranslateCWMetricToEMF(cwMetricLists []*CWMetrics) []*LogEvent {
```
Are we using `TranslateCWMetricToEMF` anywhere?

kohrapha commented on exporter/awsemfexporter/pusher_test.go:
```diff
@@ -79,7 +79,7 @@ func TestLogEventBatch_sortLogEvents(t *testing.T) {
 			int64(timestamp),
 			fmt.Sprintf("message%v", timestamp))
-		fmt.Printf("logEvents[%d].Timetsmap=%d.\n", i, timestamp)
+		fmt.Printf("logEvents[%d].Timestamp=%d.\n", i, timestamp)
```
nice :+1:

kohrapha commented on exporter/awsemfexporter/metric_translator_test.go:
```diff
@@ -1323,36 +1333,179 @@ func BenchmarkTranslateOtToCWMetricWithNamespace(b *testing.B) {
+	fields_1 := make(map[string]interface{})
+	fields_1["nginx_ingress_controller_nginx_process_connections"] = 0
+	fields_1["controller_pod"] = "my-nginx-ingress-nginx-controller-bbf548c86-wl84b"
+	fields_1["kubernetes_node"] = "ip-192-168-17-95.us-west-2.compute.internal"
```
nit: I would suggest using dummy values here for test cases, like `field1` or `a` vs using these "real-world" specific values.

kohrapha reviewed (changes requested):
Gave another look through, more detailed round of feedback. Overall, good job, but I'm still a bit unsure of the positioning of the batching logic and the design.

mxiamxia commented on exporter/awsemfexporter/emf_exporter.go:
```diff
@@ -174,21 +174,23 @@ func generateLogEventFromMetric(metric pdata.Metrics, dimensionRollupOption stri
 			namespace = cwm[0].Measurements[0].Namespace
 		}
 		// append all datapoint metrics in the request into CWMetric list
 		cwMetricLists = append(cwMetricLists, cwm...)
```
These piece of code doesn't looks clean. do we still use `cwMetricLists`? I think we need to reconsider where to do the batch logic

mxiamxia commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -61,6 +62,15 @@ type CWMetrics struct {
 
+// GroupedCWMetric defines 
+type GroupedCWMetric struct {
+	Namespace    string
+	Metrics      map[string]interface{}
+	Timestamp    int64
+	Dimensions   map[string]interface{}
+	MetricUnits  map[string]string
```
how do we make sure the metrics match the right dimensions? we have two separate maps for metrics and dimensions.

bryanlou replied:
fixed

hdj630 commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -55,15 +54,22 @@ type rateState struct {
+type GroupedMetric struct {
+	Namespace    string
 	Timestamp    int64
-	Fields       map[string]interface{}
```
In our latest discussion, Labels should be defined as map[string]string

bryanlou replied:
Right now defined as interface{}, will change accordingly

hdj630 commented on exporter/awsemfexporter/factory.go:
```diff
@@ -61,7 +61,7 @@ func createDefaultConfig() configmodels.Exporter {
 func createMetricsExporter(_ context.Context,
 	params component.ExporterCreateParams,
 	config configmodels.Exporter) (component.MetricsExporter, error) {
-
```
remove this

bryanlou replied:
ok

hdj630 commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -118,13 +124,8 @@ func (dps DoubleHistogramDataPointSlice) At(i int) DataPoint {
-	var instrumentationLibName string
-
+func GetNamespaceFromMetric(rm *pdata.ResourceMetrics) (string) {
+	var namespace string
```
What's the namespace for nginx workload? 
We need allow configurable namespace.

bryanlou replied:
This is just used to validate namespace check for empty or nil string

hdj630 commented on exporter/awsemfexporter/emf_exporter.go:
```diff
@@ -171,27 +171,11 @@ func (emf *emfExporter) Start(ctx context.Context, host component.Host) error {
-
-	return TranslateCWMetricToEMF(cwMetricLists), totalDroppedMetrics, namespace
+	groupedMetricMap := make(map[string]*GroupedMetric)
+	TranslateOtToGroupedMetric(metric, groupedMetricMap)
```
Since the parameter `groupedMetricMap` is actually for `return value` purpose. You can make it as `return value` of `TranslateOtToGroupedMetric`

bryanlou replied:
ok

hdj630 commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -140,48 +141,116 @@ func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, config *Config) ([]*CWMetr
-			cwMetrics := getCWMetrics(&metric, namespace, instrumentationLibName, config)
-			cwMetricList = append(cwMetricList, cwMetrics...)
+
+			metrics := ilm.Metrics()
```
`instrumentationLibName` is also need to be part of key for batching

bryanlou replied:
Ok, I included it in the `buildGroupedMetric` function

hdj630 commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -140,48 +141,116 @@ func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, config *Config) ([]*CWMetr
+				}
+
+				if dps.Len() == 0 {
+					return
```
Need to process other metrics instead of return from this function.

bryanlou replied:
changed to `continue`

hdj630 commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -140,48 +141,116 @@ func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, config *Config) ([]*CWMetr
+
+					labelSlice := make([]string, 0, dp.LabelsMap().Len())
+					dp.LabelsMap().ForEach(func(k string, v string) {
+						labelSlice = append(labelSlice, k)
```
Both labelkey and labelvalue matter when compose the batching key.

bryanlou replied:
done!

hdj630 commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -193,87 +262,72 @@ func TranslateCWMetricToEMF(cwMetricLists []*CWMetrics) []*LogEvent {
-		)
-		return
+	if metricVal == nil {
+		return nil
```
Do you need to check the return value `nil` in the function where call this function?

bryanlou replied:
yes I think so

hdj630 commented on exporter/awsemfexporter/metric_translator.go:
```diff
@@ -140,48 +141,116 @@ func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, config *Config) ([]*CWMetr
+						labelSlice = append(labelSlice, k)
+					})
+					sort.Strings(labelSlice)
+					key := strings.Join(labelSlice, "")
```
Need other field key and value to be batching key. i.e `Namespace`, `instrumentationLibName`, `Timestamp`

bryanlou replied:
I added `Namespace` value as part of the batching key

kohrapha commented:
Closing this in favour of the new PR #62
